### DISCUSSION TITLE
about improve

### DISCUSSION THREAD

**Discussion by lo-th** [jrouwe/JoltPhysics.js#13] [Q&A]
Perf boost on last version is great !! 
( the build in Examples/js/  is not the good one )
I start deep test on PHY, i tell you soon what need 

this engine have raycast ?

**Comment by jrouwe** [comment]
Hello,

> the build in Examples/js/ is not the good one

There is no build in this repository anymore. The build was moved to an npm package.

> this engine have raycast ?

Yes, but I didn't expose it in the js interface yet.

**Comment by lo-th** [comment]
keep the build maybe in build folder ( npm package is great but have it directly with github is usefull )
this jolt-physics.wasm-compat.js is a very good idea.
js and wasm in same module file is cool.

cool for ray 
great job

**Comment by lo-th** [comment]
one question on shape creation what is 0.05 
new Jolt.BoxShape( this.v.set( 1, 1, 1), 0.05, null );

density or mass ?

**Comment by lo-th** [comment]
![jolt](https://github.com/jrouwe/JoltPhysics.js/assets/1553208/e244bad6-63bd-47c9-99e2-1b575a7fce34)
yes perf is very good 1000 object at 60 fps

**Comment by jrouwe** [comment]
> keep the build maybe in build folder ( npm package is great but have it directly with github is usefull )

Putting binaries in git really bloats the repository, so that's why it is hosted on an external site

> new Jolt.BoxShape( this.v.set( 1, 1, 1), 0.05, null );

The interface is defined here:

https://github.com/jrouwe/JoltPhysics.js/blob/main/JoltJS.idl

And it is mostly the same as the C++ interface:

https://jrouwe.github.io/JoltPhysics/class_box_shape.html#a4ccbcc1c06232059f4b8711cb54d21f8

**Comment by lo-th** [comment]
thanks 
<2 Mb is not big 
physx is 2.8 + 4.8 Mb / havok is 1.98 Mb

i will send link to demo soon
perf is near this engine !

**Comment by lo-th** [comment]
It's an unusual method, but I'm starting to understand.
it's better to use Jolt.BoxShape or Jolt.BoxShapeSettings ? for shape definition.
 
i don't find a way to set density or mass ? 

body and world reset work fine.

now i have working compound ( inConvexRadius should be near 0 on shapeSetting )
one need is DOF constraint

**Comment by jrouwe** [comment]
> it's better to use Jolt.BoxShape or Jolt.BoxShapeSettings ? for shape definition.

Both are fine. More complex shapes only have XXXSettings classes to configure them.

> i don't find a way to set density or mass ?

Either you call SetDensity on the shape or you set BodyCreationSettings.mOverrideMassProperties and mMassPropertiesOverride.mMass.

**Comment by lo-th** [comment]
here the [demo](https://lo-th.github.io/phy/index.html?E=w_jolt#simple) :)
note: few demo works for the moment

Mass or density seem not work well, i try both  ?
i have problem's with collision filter is possible to set group but not mask ( to restrict collision )

**Comment by jrouwe** [comment]
> Mass or density seem not work well, i try both ?

They're mutually exclusive, either you set mass OR you set density. As far as I know it should be working (have been using it for years).

> i have problem's with collision filter is possible to set group but not mask ( to restrict collision )

JoltPhysics supports this but it is not exposed to JoltPhysics.js. It would involve replacing

https://github.com/jrouwe/JoltPhysics.js/blob/0f3538a7a9615cbdbafa452f12997e3ea6a9fd55/JoltJS.h#L81-L171

with JavaScript callbacks (something I don't have time for at the moment).

**Reply by lo-th** [reply]
ok i will do more test 

i see collision filter look complex in jolt
generally only need two number group and mask

**Comment by jrouwe** [comment]
Looking at your code:

https://github.com/lo-th/phy/blob/1303a3ad9c9aa2bdff4e81b4c252911755fc342b/src/_jolt/Body.js#L303

You're forgetting to set mOverrideMassProperties to CalculateInertia, this means that the mMass you're setting here won't be used.

**Comment by lo-th** [comment]
ok thanks i add bcs.mOverrideMassProperties = 1
i use my own volume calculation so only set mass is perfect.

i try AddImpulse and AddForce but don't work has expected ?

also i see friction and restitution is set to high first
in majority of engine is lower first or product

i also see a problem's look "[random](https://lo-th.github.io/phy/index.html?E=w_jolt#random)" demo 
after long time of running i see bug like body list is full 
for delete i use 
root.bodyInterface.RemoveBody( b.GetID() ); 
root.bodyInterface.DestroyBody( b.GetID() ); 

i think i find bug if i set friction to 0 is not apply

**Comment by jrouwe** [comment]
> i try AddImpulse and AddForce but don't work has expected ?

The default mass of objects in Jolt is probably much higher than other engines. If you didn't set the mass properly, you won't see much happening. Try increasing the magnitude by a factor 1000 to see if it will then work (and if it does you still have an issue with mass).

> in majority of engine is lower first or product

The friction/restitution functions can be configured, but I did not expose this to JavaScript. If you have C++ experience then you can add it and send me a PR (or you have to wait for me to have time to look into these things).

https://github.com/jrouwe/JoltPhysics/blob/0bbfc1518dfb5da4d76c96e2e351a5f10df805cd/Jolt/Physics/PhysicsSystem.h#L54-L62

> root.bodyInterface.RemoveBody( b.GetID() );
> root.bodyInterface.DestroyBody( b.GetID() );

This should work, are you sure b.GetID() is a valid ID?

> i think i find bug if i set friction to 0 is not apply

Not sure what you mean with this, but if you set the friction of one object to 0 then the total friction is zero according to the default formula: sqrt(friction1 * friction2)

**Reply by lo-th** [reply]
default scale for all engine is 1 unit = meter and 1 mass in kg.
compared to the other engine it seems correct. 
i will try force with bigger value, physx also use bigger value

yes i see they are function to configure friction/restitution ratio like havok is not very important for the moment

yes i'm sure is valid id, Maybe i think internal id in jolt is not free and always increase to the limite

for friction 0 is a code error maybe at 0.01 object slide like soap but 0 no more slide
This often happens to me when I use if(value) v = value instead of if(value!==undefined) v = value;
but is maybe another problem

**Reply by jrouwe** [reply]
I created two tests:

https://jrouwe.nl/jolt/examples/add_remove_bodies.html

Which seems to run forever (and I can see that it doesn't leak if I log the body IDs).

And:

https://jrouwe.nl/jolt/examples/friction.html

Which shows that friction = 0 works just fine for me. It also uses AddImpulse without any issues.

**Reply by jrouwe** [reply]
Note though that if you call AddImpulse on a sleeping object, nothing will happen. You need to activate the body yourself.

**Reply by lo-th** [reply]
ok great 
i will look on my side

**Comment by jrouwe** [comment]
FYI the SixDOF joint has been exposed (and motor support for all joints too)

**Comment by jrouwe** [comment]
Ray casting has been exposed too b.t.w.

**Comment by lo-th** [comment]
top

**Comment by lo-th** [comment]
hi collision work great 
for ray i don't see how run 
i have a ray
ray = new Jolt.RayCast();
i set mDirection and mOrigin
but how perform raycast and get return value ? if you have idea

**Reply by jrouwe** [reply]
Sorry that was indeed not very clear and I see I didn't properly test this myself either. There's now a new version and a demo:

https://jrouwe.github.io/JoltPhysics.js/ray_cast.html

**Reply by lo-th** [reply]
ok now ray work 
in other engine ray return also hit normal ( normal of hit face )
do you think is possible in Jolt ?

**Comment by lo-th** [comment]
they are something you should add if possible
for Jolt.Vec3 and Jolt.Quat they are no clear function to set value.

no need create new one each time so i make this functiion

```
Jolt.Vec3.prototype.set = function ( x,y,z ){
	this.SetX( x );
	this.SetY( y );
	this.SetZ( z );
	return this;
}
```

for Quat  this.SetZ( z ); not define, so i use constructor 
but is change on every build :)
```
Jolt.Quat.prototype.set = function ( x,y,z,w ){
	this.EJ(x,y,z,w);
	return this;
}
```

**Comment by jrouwe** [comment]
* I added the hit normal [to the example](https://jrouwe.github.io/JoltPhysics.js/ray_cast.html) (green line)
* I added a Set function to both Quat and Vec3

**Reply by lo-th** [reply]
Hi i'm near good joint implementation 
i have some question
on creation joint  need jc = new Jolt.HingeConstraintSettings()
and on creation joint = jc.Create( b1, b2 );

but all setting like limite motor ... are on jc not on joint 
so i can't change setting after joint creation ? ( i try change jc but not work )
maybe need add reference of jc in joint 

also i see you use World position for joint 
all other engine use local position for joint position on body
that a bit confusing but why not.

i have to test more for the moment joint is not strong 
i will send demo soon 

great add for vehicle and bike !!

**Reply by jrouwe** [reply]
> but all setting like limite motor ... are on jc not on joint

Yes they are:

https://github.com/jrouwe/JoltPhysics.js/blob/40211f90ed9636dda39ead25dce55b0adaeabcfe/JoltJS.idl#L901

You do need to cast the joint using ```Jolt.castObject(joint, Jolt.HingeConstraint)``` though.

> also i see you use World position for joint

You can use local space positions (or more precisely positions relative to center of mass) too:

https://github.com/jrouwe/JoltPhysics.js/blob/40211f90ed9636dda39ead25dce55b0adaeabcfe/JoltJS.idl#L874

**Reply by lo-th** [reply]
ho i don't see this one great thank's
just have to set jc.mSpace = 0; default is 1 world space

**Reply by lo-th** [reply]
ok so my joint = jc.Create( b1, b2 );
are joint = Jolt.castObject( jc.Create( b1, b2) , Jolt.type )
and jc can be detroy
top thank's

**Reply by lo-th** [reply]
ok demo here 
https://lo-th.github.io/phy/index.html?E=w_jolt#joint
now just have to add all option for each joint  ( only hinge is set now )
 ❤️
